src/VisualStudio/Core/Def/Venus/ContainedLanguage.cs

```diff
@@ -115,13 +128,21 @@ public static string GetFilePathFromHierarchyAndItemId(IVsHierarchy hierarchy, u
             var bufferTagAggregatorFactory = ComponentModel.GetService<IBufferTagAggregatorFactoryService>();
             _bufferTagAggregator = bufferTagAggregatorFactory.CreateTagAggregator<ITag>(SubjectBuffer);
 
+            // TODO: Can contained documents be linked or shared?
+            this.DataBuffer.Changed += OnDataBufferChanged;
+        }
+
+        private void InitializeContainedDocument(ProjectId projectId, string filePath, AbstractFormattingRule? vbHelperFormattingRule = null)
+        {
```

Comment:
unlcear to me why this code can't just be in the constructor.

Reply:
well... i think we can still do this without duplication.  just have all the constructors defer to one main constructor and do this work in that constructor.  :)

Reply:
Could you decorate the Initialize method with the MemberNotNullAttribute if you didn't want to do the work in a ctor?

Reply:
Yeah I'll do that. I had not done it before because I wanted to avoid the super long parameter list and because we assume that if some parameters are present then so are others, but I'll clean things up.